mantidproject/mantid
# Fix crash in DiscusMultipleScatteringCorrection

**Description of work.**

Some long calculations I was running on PEARL data sets crashed due to an exception being raised in the part of the code that determines the new track direction in between scattering events. This is in the method DiscusMultipleScatteringCorrection::updateTrackDirection
Main problem was that output V3D vector wasn't always of size 1 due to a floating point rounding problem when the input vector was close to vertical (vz~1). This raised an exception in Track::reset.
I've changed `sqrt(1-vz*vz)` to `sqrt(vx*vx + vy*vy)` to avoid subtracting two numbers that are v close together.
Also extend scenarios covered by special case code to cover track that is going vertically down as well as vertically up (ie for vz=-1 as well as vz=1)

**To test:**

I only reproduced the crash in a calculation that took a couple of hours to run. So don't have a nice script that shows the problem before and after. So suggest just doing a code review. The algorithm is covered by various tests on the calculation results in the unit test so the Jenkins checks should ensure these changes haven't changed the results.

Fixes #32988.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
